flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/functions/FromElementsGeneratorFunction.java

```diff
@@ -0,0 +1,211 @@
+ * object transport using Java serialization will not be affected by the serializability of the
+ * elements.
+ *
+ * <p><b>NOTE:</b> This source has a parallelism of 1.
+ *
+ * @param <OUT> The type of elements returned by this function.
+ */
+@Internal
+public class FromElementsGeneratorFunction<OUT>
+        implements GeneratorFunction<Long, OUT>, OutputTypeConfigurable<OUT> {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final Logger LOG = LoggerFactory.getLogger(FromElementsGeneratorFunction.class);
+
+    /** The (de)serializer to be used for the data elements. */
+    private @Nullable TypeSerializer<OUT> serializer;
+
+    /** The actual data elements, in serialized form. */
+    private byte[] elementsSerialized;
+
+    /** The number of elements emitted already. */
+    private int numElementsEmitted;
+
+    private final transient Iterable<OUT> elements;
+    private transient DataInputView input;
+
+    public FromElementsGeneratorFunction(TypeSerializer<OUT> serializer, OUT... elements)
+            throws IOException {
+        this(serializer, Arrays.asList(elements));
+    }
+
+    public FromElementsGeneratorFunction(TypeSerializer<OUT> serializer, Iterable<OUT> elements)
+            throws IOException {
+        this.serializer = Preconditions.checkNotNull(serializer);
+        this.elements = elements;
+        serializeElements();
+    }
+
+    @SafeVarargs
+    public FromElementsGeneratorFunction(OUT... elements) {
+        this(Arrays.asList(elements));
+    }
+
+    public FromElementsGeneratorFunction(Iterable<OUT> elements) {
+        this.serializer = null;
+        this.elements = elements;
+        checkIterable(elements, Object.class);
+    }
+
+    @VisibleForTesting
+    @Nullable
+    public TypeSerializer<OUT> getSerializer() {
+        return serializer;
+    }
+
+    private void serializeElements() throws IOException {
+        Preconditions.checkState(serializer != null, "serializer not set");
+        LOG.info("Serializing elements using  " + serializer);
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        DataOutputViewStreamWrapper wrapper = new DataOutputViewStreamWrapper(baos);
+
+        try {
+            for (OUT element : elements) {
+                serializer.serialize(element, wrapper);
+            }
+        } catch (Exception e) {
+            throw new IOException("Serializing the source elements failed: " + e.getMessage(), e);
+        }
+        this.elementsSerialized = baos.toByteArray();
+    }
+
+    @Override
+    public void open(SourceReaderContext readerContext) throws Exception {
+        ByteArrayInputStream bais = new ByteArrayInputStream(elementsSerialized);
+        this.input = new DataInputViewStreamWrapper(bais);
+    }
+
+    @Override
+    public OUT map(Long nextIndex) throws Exception {
+        // Move iterator to the required position in case of failure recovery
```

Comment:
Hmmm, indeed. This would work not only in case of failure recovery, but when the underlying NumberSequenceSource splits the range between the parallel instances. It is a bit wasteful to ship the whole data collection to each parallel source task, but hey, not that we expect millions of records there. Only the question from the above thread regarding the intention of the non parallel original implementation remains.